# Hazem Zaky
# Programming Exercise \#1
---

# Preamble

In [677]:
# optional code cell when using Google Colab with Google Drive

# remove the docstring comment block below in order to mount Google Drive
'''
# mount Google Drive in Google Colab
from google.colab import drive
drive.mount('/content/drive')

# change directory using the magic command %cd
### replace [MY PATH] below with your own path in Google Drive ###
### %cd /content/drive/My\ Drive/[MY PATH] ###
'''

"\n# mount Google Drive in Google Colab\nfrom google.colab import drive\ndrive.mount('/content/drive')\n\n# change directory using the magic command %cd\n### replace [MY PATH] below with your own path in Google Drive ###\n### %cd /content/drive/My\\ Drive/[MY PATH] ###\n"

In [678]:
# import relevant Python libraries

import numpy as np
import pandas as pd
from scipy import stats as sps
from IPython.display import display, Latex

# **1. Fetal Health Classification Dataset**

## **Clean Dataset**

### Problem 1.1

In [679]:
# load the clean dataset csv file into a pandas dataframe

fetal_df = pd.read_csv('data/fetal_health_dataset_clean.csv') 

#### (a)

This is a supervised machine learning task because we are given the output of each data value as whether the fetus is normal, suspect, or pathological.

#### (b)

In [680]:
print('Axes: ', fetal_df.axes)
print('\n\n')
print('Dtypes: ', fetal_df.dtypes)

Axes:  [RangeIndex(start=0, stop=2126, step=1), Index(['baseline value', 'accelerations', 'fetal_movement',
       'uterine_contractions', 'light_decelerations', 'severe_decelerations',
       'prolongued_decelerations', 'abnormal_short_term_variability',
       'mean_value_of_short_term_variability',
       'percentage_of_time_with_abnormal_long_term_variability',
       'mean_value_of_long_term_variability', 'histogram_width',
       'histogram_min', 'histogram_max', 'histogram_number_of_peaks',
       'histogram_number_of_zeroes', 'histogram_mode', 'histogram_mean',
       'histogram_median', 'histogram_variance', 'histogram_tendency',
       'fetal_health'],
      dtype='object')]



Dtypes:  baseline value                                              int64
accelerations                                             float64
fetal_movement                                            float64
uterine_contractions                                      float64
light_decelerations           

#### (c)

In [681]:
fetal_df.head(10)

baseline value  accelerations  fetal_movement  uterine_contractions  \
0             120          0.000             0.0                 0.000   
1             132          0.006             0.0                 0.006   
2             133          0.003             0.0                 0.008   
3             134          0.003             0.0                 0.008   
4             132          0.007             0.0                 0.008   
5             134          0.001             0.0                 0.010   
6             134          0.001             0.0                 0.013   
7             122          0.000             0.0                 0.000   
8             122          0.000             0.0                 0.002   
9             122          0.000             0.0                 0.003   

   light_decelerations  severe_decelerations  prolongued_decelerations  \
0                0.000                   0.0                     0.000   
1                0.003                   0.0                     0.000   
2                0.003                   0.0                     0.000   
3                0.003                   0.0                     0.000   
4                0.000                   0.0                     0.000   
5                0.009                   0.0                     0.002   
6                0.008                   0.0                     0.003   
7                0.000                   0.0                     0.000   
8                0.000                   0.0                     0.000   
9                0.000                   0.0                     0.000   

   abnormal_short_term_variability  mean_value_of_short_term_variability  \
0                               73                                   0.5   
1                               17                                   2.1   
2                               16                                   2.1   
3                               16                                   2.4   
4                               16                                   2.4   
5                               26                                   5.9   
6                               29                                   6.3   
7                               83                                   0.5   
8                               84                                   0.5   
9                               86                                   0.3   

   percentage_of_time_with_abnormal_long_term_variability  ...  histogram_min  \
0                                                 43       ...             62   
1                                                  0       ...             68   
2                                                  0       ...             68   
3                                                  0       ...             53   
4                                                  0       ...             53   
5                                                  0       ...             50   
6                                                  0       ...             50   
7                                                  6       ...             62   
8                                                  5       ...             62   
9                                                  6       ...             62   

   histogram_max  histogram_number_of_peaks  histogram_number_of_zeroes  \
0            126                          2                           0   
1            198                          6                           1   
2            198                          5                           1   
3            170                         11                           0   
4            170                          9                           0   
5            200                          5                           3   
6            200                          6                           3   
7            130                          0 

#### (d)

In [682]:
fetal_df.shape

(2126, 22)

#### (e)

Each row is called a data sample.

#### (f)

There are 2126 total samples.

#### (g)

21 independent variables.
baseline value
accelerations
fetal_movement
uterine_contractions
light_decelerations
severe_decelerations
prolongued_decelerations
abnormal_short_term_variability
mean_value_of_short_term_variability
percentage_of_time_with_abnormal_long_term_variability
mean_value_of_long_term_variability
histogram_width
histogram_min
histogram_max
histogram_number_of_peaks
histogram_number_of_zeroes
histogram_mode
histogram_mean
histogram_median
histogram_variance
histogram_tendency


#### (h)

1 Dependent variable
fetal_health

#### (i)

n = 2126, p = 21


#### (j)

n = 2126, m = 1

#### (k)

Some of the variables in the dataset are preprocessed variables, and some are raw variables. This is because some of the variables, such as histogram_tendency, histogram_variance, etc. Are based on the general data obtained from the histograms the raw data produced, whereas some variables such as accelerations, fetal_movement, etc. are variables that are obtained directly from the raw data.

#### (l)

In [683]:
for column in fetal_df.columns:
    unique_values = pd.unique(fetal_df[column])
    print(f"Unique values in '{column}': {sorted(unique_values)}\n")

Unique values in 'baseline value': [106, 110, 112, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 154, 156, 157, 158, 159, 160]

Unique values in 'accelerations': [0.0, 0.001, 0.002, 0.003, 0.004, 0.005, 0.006, 0.007, 0.008, 0.009, 0.01, 0.011, 0.012, 0.013, 0.014, 0.015, 0.016, 0.017, 0.018, 0.019]

Unique values in 'fetal_movement': [0.0, 0.001, 0.002, 0.003, 0.004, 0.005, 0.006, 0.007, 0.008, 0.009, 0.01, 0.011, 0.012, 0.013, 0.014, 0.015, 0.016, 0.017, 0.018, 0.019, 0.02, 0.021, 0.022, 0.023, 0.024, 0.025, 0.026, 0.027, 0.028, 0.029, 0.03, 0.031, 0.032, 0.033, 0.035, 0.036, 0.037, 0.038, 0.04, 0.041, 0.043, 0.045, 0.047, 0.048, 0.049, 0.05, 0.051, 0.052, 0.053, 0.054, 0.055, 0.058, 0.06, 0.063, 0.065, 0.071, 0.072, 0.079, 0.084, 0.085, 0.088, 0.089, 0.091, 0.092, 0.099, 0.103, 0.107, 0.108, 0.109, 0.112, 0.115, 0.135, 0.139, 0.157, 0.18

histogram_tendency
fetal_health

These are the only categorical variables because the rest of the variables have continuous values that do not have specific bins.


#### (m)

The histogram_tendency variable goes by nominal encoding, where the numbers don't have a specific meaning. The fetal_health variable goes by ordinal encoding because the unique values represent categories of fetal health, with 1 being the healthiest and 3 being the least healthy.

#### (n)

In [684]:
normal = fetal_df[fetal_df['fetal_health'] == 1].count().values[0]
suspect = fetal_df[fetal_df['fetal_health'] == 2].count().values[0]
pathological = fetal_df[fetal_df['fetal_health'] == 3].count().values[0]
print("Normal: ",normal)
print("Suspect: ",suspect)
print("Pathological: ",pathological)


Normal:  1655
Suspect:  295
Pathological:  176


## **Dirty Dataset**

In [685]:
# load the dirty dataset csv file into a pandas dataframe

fetal_dirty_df = pd.read_csv('data/fetal_health_dataset_dirty.csv') 

### Problem 1.2

#### (a)

In [686]:
fetal_dirty_df.isna().sum()

baseline value                                            47
accelerations                                              0
fetal_movement                                             0
uterine_contractions                                       0
light_decelerations                                        0
severe_decelerations                                       0
prolongued_decelerations                                   0
abnormal_short_term_variability                            0
mean_value_of_short_term_variability                       0
percentage_of_time_with_abnormal_long_term_variability    10
mean_value_of_long_term_variability                        0
histogram_width                                            0
histogram_min                                              0
histogram_max                                             24
histogram_number_of_peaks                                  0
histogram_number_of_zeroes                                 0
histogram_mode          

#### (b)

In [687]:
fetal_dirty_df.columns[fetal_dirty_df.isna().sum() != 0].tolist()

['baseline value',
 'percentage_of_time_with_abnormal_long_term_variability',
 'histogram_max',
 'histogram_mode']

#### (c)

In [688]:
fetal_dirty_df[fetal_dirty_df.isna().any(axis=1)].shape[0]

86

### Problem 1.3

In [689]:
fetal_dirty_df.loc[fetal_dirty_df['baseline value'] < 0,'baseline value'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['baseline value'] > 240,'baseline value'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['accelerations'] < 0,'accelerations'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['accelerations'] > 0.03,'accelerations'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['fetal_movement'] < 0,'fetal_movement'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['fetal_movement'] > .15,'fetal_movement'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['uterine_contractions'] < 0,'uterine_contractions'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['uterine_contractions'] > 0.02,'uterine_contractions'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['light_decelerations'] < 0,'light_decelerations'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['light_decelerations'] > 0.02,'light_decelerations'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['severe_decelerations'] < 0,'severe_decelerations'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['severe_decelerations'] > 0.02,'severe_decelerations'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['prolongued_decelerations'] < 0,'prolongued_decelerations'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['prolongued_decelerations'] > 0.02,'prolongued_decelerations'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['abnormal_short_term_variability'] < 0,'abnormal_short_term_variability'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['abnormal_short_term_variability'] > 100,'abnormal_short_term_variability'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['mean_value_of_short_term_variability'] < 0,'mean_value_of_short_term_variability'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['mean_value_of_short_term_variability'] > 10,'mean_value_of_short_term_variability'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['percentage_of_time_with_abnormal_long_term_variability'] < 0,'percentage_of_time_with_abnormal_long_term_variability'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['percentage_of_time_with_abnormal_long_term_variability'] > 100,'percentage_of_time_with_abnormal_long_term_variability'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['mean_value_of_long_term_variability'] < 0,'mean_value_of_long_term_variability'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['mean_value_of_long_term_variability'] > 100,'mean_value_of_long_term_variability'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_width'] < 0,'histogram_width'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_width'] > 2126,'histogram_width'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_min'] < 0,'histogram_min'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_min'] > 200,'histogram_min'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_max'] < 50,'histogram_max'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_max'] > 300,'histogram_max'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_number_of_peaks'] < 0,'histogram_number_of_peaks'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_number_of_peaks'] > 100,'histogram_number_of_peaks'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_number_of_zeroes'] < 0,'histogram_number_of_zeroes'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_number_of_zeroes'] > 50,'histogram_number_of_zeroes'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_mode'] < 0,'histogram_mode'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_mode'] > 250,'histogram_mode'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_mean'] < 0,'histogram_mean'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_mean'] > 250,'histogram_mean'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_median'] < 0,'histogram_median'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_median'] > 250,'histogram_median'] = np.nan

fetal_dirty_df.loc[fetal_dirty_df['histogram_variance'] < 0,'histogram_variance'] = np.nan
fetal_dirty_df.loc[fetal_dirty_df['histogram_variance'] > 350,'histogram_variance'] = np.nan

fetal_dirty_df.loc[~fetal_dirty_df['histogram_tendency'].isin([1, 0, -1]), 'fetal_health'] = np.nan

fetal_dirty_df.loc[~fetal_dirty_df['fetal_health'].isin([1, 2, 3]), 'fetal_health'] = np.nan
fetal_dirty_df.loc[91:100]


baseline value  accelerations  fetal_movement  uterine_contractions  \
91            148.0          0.000           0.000                 0.000   
92            148.0          0.000           0.003                 0.000   
93              NaN          0.000           0.000                 0.000   
94            148.0          0.000           0.005                 0.002   
95            148.0          0.000           0.000                 0.000   
96            148.0          0.000           0.000                 0.000   
97            148.0          0.000           0.001                 0.000   
98            148.0          0.000           0.001                 0.000   
99            125.0          0.010           0.007                 0.002   
100           125.0          0.005           0.000                 0.002   

     light_decelerations  severe_decelerations  prolongued_decelerations  \
91                 0.000                   0.0                       0.0   
92                 0.000                   0.0                       0.0   
93                 0.000                   0.0                       0.0   
94                 0.000                   0.0                       0.0   
95                 0.000                   0.0                       0.0   
96                 0.000                   0.0                       0.0   
97                 0.001                   0.0                       0.0   
98                 0.001                   0.0                       0.0   
99                 0.000                   0.0                       0.0   
100                0.003                   0.0                       0.0   

     abnormal_short_term_variability  mean_value_of_short_term_variability  \
91                              67.0                                   0.4   
92                              70.0                                   0.3   
93                              68.0                                   0.3   
94                              72.0                                   0.3   
95                              58.0                                   0.5   
96                              63.0                                   0.4   
97                              60.0                                   0.5   
98                              61.0                                   0.5   
99                              19.0                                   1.9   
100                             25.0                                   1.7   

     percentage_of_time_with_abnormal_long_term_variability  ...  \
91                                                67.0       ...   
92                                                68.0       ...   
93                                                75.0       ...   
94                                                74.0       ...   
95                                                 7.0       ...   
96                                                30.0       ...   
97                                                49.0       ...   
98                                                39.0       ...   
99                                                 0.0       ...   
100                                                6.0       ...   

     histogram_min  histogram_max  histogram_number_of_peaks  \
91           139.0          154.0                        4.0   
92           136.0          153.0                        2.0   
93           128.0            NaN                        3.0   
94           136.0          152.0                        0.0   
95           135.0          156.0                        1.0   
96           121.0          156.0                        3.0   
97            78.0          161.0                        3.0   
98           130.0          161.0                        2.0   
99           101.0          197.0                        8.0   
100           72.0          165.0                        3.0   

     histogram_num

The baseline value ranges from about 40 bpm to 180 bpm, which includes whether the baby has barychardia or tachycardia, which indicates a lower than normal or higher than normal heartrate in babies. Given that accelerations have an average of about 34 per hour, a safe assumption would be that an upper bound would be 100 per hour, which roughly translates to about 0.03 per second. Average fetal movement is can range from 10-30 per hour, but can possibly hit several hundred movements per hour during periods of extremely high activity. With this, I limit the fetal movement to 500 per hour, and consider anything above it to be infeasible. The number of contractions can increase to almost 60 contractions per hour, making a good upper bound of about 0.02 per minute. All of the decelerations range up to about 0.005 per second, so a value of 0.2 is a strict upper bound. For the mean value of short term and long term variability, my upper bounds were about 2-3x the upper bounds discovered through research. The percentages all range to 100. Any lower bound not specifically mentioned can be assumed to be non-negative values.


### Problem 1.4

In [690]:
curr_columns = fetal_dirty_df.columns
for col in curr_columns:
    if (col == 'fetal_health' or col == 'histogram_tendency'):
        continue
    curr_mean = fetal_dirty_df[col].mean(skipna=True)
    curr_std = fetal_dirty_df[col].std(skipna=True)
    if col == "severe_decelerations":
        print(curr_mean)
        print(curr_std)
    fetal_dirty_df[col] = (fetal_dirty_df[col]-curr_mean)/curr_std

fetal_dirty_df.head(20)

3.292568203198495e-06
5.7299838918067855e-05


baseline value  accelerations  fetal_movement  uterine_contractions  \
0        -1.350760      -0.822195       -0.329965             -1.482116   
1        -0.131036       0.729961       -0.329965              0.554496   
2        -0.029392      -0.046117       -0.329965              1.233367   
3         0.072252      -0.046117       -0.329965              1.233367   
4        -0.131036       0.988654       -0.329965              1.233367   
5         0.072252      -0.563502       -0.329965              1.912238   
6         0.072252      -0.563502       -0.329965              2.930544   
7        -1.147472      -0.822195       -0.329965             -1.482116   
8        -1.147472      -0.822195       -0.329965             -0.803245   
9        -1.147472      -0.822195       -0.329965             -0.463810   
10        1.800194      -0.822195       -0.329965             -1.142681   
11        1.698550      -0.822195       -0.329965             -1.142681   
12       -0.232679       0.471268        5.390291              1.233367   
13       -0.232679       1.506039             NaN              0.554496   
14       -0.334323       0.729961             NaN             -0.124374   
15       -0.334323       0.729961             NaN             -0.124374   
16       -0.334323       0.729961             NaN              0.215061   
17       -0.232679      -0.304810             NaN             -0.463810   
18       -0.334323      -0.046117             NaN              0.554496   
19       -0.334323       0.471268             NaN              0.215061   

    light_decelerations  severe_decelerations  prolongued_decelerations  \
0             -0.638287             -0.057462                 -0.268691   
1              0.375155             -0.057462                 -0.268691   
2              0.375155             -0.057462                 -0.268691   
3              0.375155             -0.057462                 -0.268691   
4             -0.638287             -0.057462                 -0.268691   
5              2.402039             -0.057462                  3.121441   
6              2.064225             -0.057462                  4.816507   
7             -0.638287             -0.057462                 -0.268691   
8             -0.638287             -0.057462                 -0.268691   
9             -0.638287             -0.057462                 -0.268691   
10            -0.300473             -0.057462                 -0.268691   
11            -0.300473             -0.057462                 -0.268691   
12             0.375155             -0.057462                 -0.268691   
13             0.037341             -0.057462                 -0.268691   
14             1.050783             -0.057462                  1.426375   
15             0.712969             -0.057462                  1.426375   
16             1.050783             -0.057462                 -0.268691   
17             1.050783             -0.057462                  3.121441   
18             0.712969             -0.057462                  1.426375   
19             0.712969             -0.057462                  1.426375   

    abnormal_short_term_variability  mean_value_of_short_term_variability  \
0                          1.512834                             -0.942873   
1                         -1.744341                              0.868636   
2                         -1.802504                              0.868636   
3                         -1.802504                              1.208294   
4                         -1.802504                              1.208294   
5                         -1.220866                              5.170971   
6                         -1.046375                              5.623848   
7                          2.094473                             -0.942873   
8                          2.152636                             -0.942873   
9                          2.268964                             -1.169312   
10   

### Problem 1.5

In [691]:
curr_columns = fetal_dirty_df.columns
for col in curr_columns:
    if (col == 'fetal_health' or col == 'histogram_tendency'):
        continue
    if (col == 'fetal_health'):
        print("error here!")
    curr_median = fetal_dirty_df[col].median(skipna=True)
    fetal_dirty_df.loc[fetal_dirty_df[col].isna(), col] = curr_median

fetal_dirty_df.head(20)

baseline value  accelerations  fetal_movement  uterine_contractions  \
0        -1.350760      -0.822195       -0.329965             -1.482116   
1        -0.131036       0.729961       -0.329965              0.554496   
2        -0.029392      -0.046117       -0.329965              1.233367   
3         0.072252      -0.046117       -0.329965              1.233367   
4        -0.131036       0.988654       -0.329965              1.233367   
5         0.072252      -0.563502       -0.329965              1.912238   
6         0.072252      -0.563502       -0.329965              2.930544   
7        -1.147472      -0.822195       -0.329965             -1.482116   
8        -1.147472      -0.822195       -0.329965             -0.803245   
9        -1.147472      -0.822195       -0.329965             -0.463810   
10        1.800194      -0.822195       -0.329965             -1.142681   
11        1.698550      -0.822195       -0.329965             -1.142681   
12       -0.232679       0.471268        5.390291              1.233367   
13       -0.232679       1.506039       -0.329965              0.554496   
14       -0.334323       0.729961       -0.329965             -0.124374   
15       -0.334323       0.729961       -0.329965             -0.124374   
16       -0.334323       0.729961       -0.329965              0.215061   
17       -0.232679      -0.304810       -0.329965             -0.463810   
18       -0.334323      -0.046117       -0.329965              0.554496   
19       -0.334323       0.471268       -0.329965              0.215061   

    light_decelerations  severe_decelerations  prolongued_decelerations  \
0             -0.638287             -0.057462                 -0.268691   
1              0.375155             -0.057462                 -0.268691   
2              0.375155             -0.057462                 -0.268691   
3              0.375155             -0.057462                 -0.268691   
4             -0.638287             -0.057462                 -0.268691   
5              2.402039             -0.057462                  3.121441   
6              2.064225             -0.057462                  4.816507   
7             -0.638287             -0.057462                 -0.268691   
8             -0.638287             -0.057462                 -0.268691   
9             -0.638287             -0.057462                 -0.268691   
10            -0.300473             -0.057462                 -0.268691   
11            -0.300473             -0.057462                 -0.268691   
12             0.375155             -0.057462                 -0.268691   
13             0.037341             -0.057462                 -0.268691   
14             1.050783             -0.057462                  1.426375   
15             0.712969             -0.057462                  1.426375   
16             1.050783             -0.057462                 -0.268691   
17             1.050783             -0.057462                  3.121441   
18             0.712969             -0.057462                  1.426375   
19             0.712969             -0.057462                  1.426375   

    abnormal_short_term_variability  mean_value_of_short_term_variability  \
0                          1.512834                             -0.942873   
1                         -1.744341                              0.868636   
2                         -1.802504                              0.868636   
3                         -1.802504                              1.208294   
4                         -1.802504                              1.208294   
5                         -1.220866                              5.170971   
6                         -1.046375                              5.623848   
7                          2.094473                             -0.942873   
8                          2.152636                             -0.942873   
9                          2.268964                             -1.169312   
10   

### Problem 1.6

In [692]:

fetal_dirty_df = pd.get_dummies(fetal_dirty_df, columns=['fetal_health'], prefix='fetal_health')
display(fetal_dirty_df)
fetal_dirty_df.to_csv('fetal_health_dataset_processed.csv', index=False)

baseline value  accelerations  fetal_movement  uterine_contractions  \
0          -1.350760      -0.822195       -0.329965             -1.482116   
1          -0.131036       0.729961       -0.329965              0.554496   
2          -0.029392      -0.046117       -0.329965              1.233367   
3           0.072252      -0.046117       -0.329965              1.233367   
4          -0.131036       0.988654       -0.329965              1.233367   
...              ...            ...             ...                   ...   
2121        0.682113      -0.822195       -0.329965              0.893932   
2122        0.682113      -0.563502       -0.329965              0.893932   
2123        0.682113      -0.563502       -0.329965              0.893932   
2124        0.682113      -0.563502       -0.329965              0.554496   
2125        0.885401      -0.304810       -0.171069              1.233367   

      light_decelerations  severe_decelerations  prolongued_decelerations  \
0               -0.638287             -0.057462                 -0.268691   
1                0.375155             -0.057462                 -0.268691   
2                0.375155             -0.057462                 -0.268691   
3                0.375155             -0.057462                 -0.268691   
4               -0.638287             -0.057462                 -0.268691   
...                   ...                   ...                       ...   
2121            -0.638287             -0.057462                 -0.268691   
2122            -0.638287             -0.057462                 -0.268691   
2123            -0.638287             -0.057462                 -0.268691   
2124            -0.638287             -0.057462                 -0.268691   
2125            -0.638287             -0.057462                 -0.268691   

      abnormal_short_term_variability  mean_value_of_short_term_variability  \
0                            1.512834                             -0.942873   
1                           -1.744341                              0.868636   
2                           -1.802504                              0.868636   
3                           -1.802504                              1.208294   
4                           -1.802504                              1.208294   
...                               ...                                   ...   
2121                         1.861817                             -1.282531   
2122                         1.803653                             -1.056093   
2123                         1.861817                             -1.056093   
2124                         1.803653                             -1.056093   
2125                         1.570998                             -1.056093   

      percentage_of_time_with_abnormal_long_term_variability  ...  \
0                                              1.808583       ...   
1                                             -0.532666       ...   
2                                             -0.532666       ...   
3                                             -0.532666       ...   
4                                             -0.532666       ...   
...                                                 ...       ...   
2121                                           0.828525       ...   
2122                                           0.665182       ...   
2123                                           0.556287       ...   
2124                                           0.937421       ...   
2125                                           1.427449       ...   

      histogram_number_of_peaks  histogram_number_of_zeroes  histogram_mode  \
0                     -0.701232                   -0.458336       -1.064163   
1                      0.654983                    0.957976        0.216611   
2                      0.315929                    0.957976        0.216611   
3                      2.350251                   -0.458336   

# **2. Heart Failure Prediction Dataset**

## Problem 2.1

In [693]:
# load the dataset csv file into a pandas dataframe

heart_df = pd.read_csv('data/heart_failure_dataset.csv')

### (a)

This algorithm is supervised machine learning because we have information on the output of each data sample. Since we have this information, we know what to classify each data value as, giving us a supervised model.

### (b)

In [694]:
print('Axes: ', heart_df.axes)
print('\n\n')
print('Dtypes: ', heart_df.dtypes)

Axes:  [RangeIndex(start=0, stop=299, step=1), Index(['age', 'anaemia', 'creatinine_phosphokinase', 'diabetes',
       'ejection_fraction', 'high_blood_pressure', 'platelets',
       'serum_creatinine', 'serum_sodium', 'sex', 'smoking', 'time',
       'DEATH_EVENT'],
      dtype='object')]



Dtypes:  age                         float64
anaemia                       int64
creatinine_phosphokinase      int64
diabetes                      int64
ejection_fraction             int64
high_blood_pressure           int64
platelets                   float64
serum_creatinine            float64
serum_sodium                  int64
sex                           int64
smoking                       int64
time                          int64
DEATH_EVENT                   int64
dtype: object


### (c)



In [695]:
heart_df.head(10)

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0  75.0        0                       582         0                 20   
1  55.0        0                      7861         0                 38   
2  65.0        0                       146         0                 20   
3  50.0        1                       111         0                 20   
4  65.0        1                       160         1                 20   
5  90.0        1                        47         0                 40   
6  75.0        1                       246         0                 15   
7  60.0        1                       315         1                 60   
8  65.0        0                       157         0                 65   
9  80.0        1                       123         0                 35   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1  265000.00               1.9           130    1   
1                    0  263358.03               1.1           136    1   
2                    0  162000.00               1.3           129    1   
3                    0  210000.00               1.9           137    1   
4                    0  327000.00               2.7           116    0   
5                    1  204000.00               2.1           132    1   
6                    0  127000.00               1.2           137    1   
7                    0  454000.00               1.1           131    1   
8                    0  263358.03               1.5           138    0   
9                    1  388000.00               9.4           133    1   

   smoking  time  DEATH_EVENT  
0        0     4            1  
1        0     6            1  
2        1     7            1  
3        0     7            1  
4        0     8            1  
5        1     8            1  
6        0    10            1  
7        1    10            1  
8        0    10            1  
9        1    10            1

### (d)

In [696]:
heart_df.shape

(299, 13)

### (e)

There are 299 samples in the dataset

### (f)

There are 12 independent variables in this dataset
age, anaemia, creatinine_phosphokinase, diabetes, ejection_fraction, high_blood_pressure, platelets, serum_creatinine, serum_sodium, sex, smoking, time

### (g)

There is 1 dependent variable: DEATH_EVENT

### (h)

n = 299, p = 12

### (i)

n = 299, m = 1

### (j)

In [697]:
for column in heart_df.columns:
    unique_values = pd.unique(heart_df[column])
    print(f"Unique values in '{column}': {unique_values}\n")

Unique values in 'age': [75.    55.    65.    50.    90.    60.    80.    62.    45.    49.
 82.    87.    70.    48.    68.    53.    95.    58.    94.    85.
 69.    72.    51.    57.    42.    41.    67.    79.    59.    44.
 63.    86.    66.    43.    46.    61.    81.    52.    64.    40.
 60.667 73.    77.    78.    54.    47.    56.   ]

Unique values in 'anaemia': [0 1]

Unique values in 'creatinine_phosphokinase': [ 582 7861  146  111  160   47  246  315  157  123   81  231  981  168
   80  379  149  125   52  128  220   63  148  112  122   60   70   23
  249  159   94  855 2656  235  124  571  127  588 1380  553  129  577
   91 3964   69  260  371   75  607  789  364 7702  318  109   68  250
  110  161  113 5882  224   92  102  203  336   76   55  280   78   84
  115   66  897  154  144  133  514   59  156   61  305  898 5209   53
  328  748 1876  936  292  369  143  754  400   96  737  358  200  248
  270 1808 1082  719  193 4540  646  281 1548  805  291  482  943  185
  13

There are 6 categorical variables: anaemia, diabetes, high_blood_pressure, sex, smoking, DEATH_EVENT. These variables are all categorical because there are only 2 options: 0 and 1. This either indicates the presence or lack thereof of these variables.

### (k)

Nominal

### (l)

In [698]:
deceased_patients = heart_df[heart_df['DEATH_EVENT'] == 1].count().values[0]
living_patients = heart_df[heart_df['DEATH_EVENT'] == 0].count().values[0]
print("Deceased Patients: ",deceased_patients)
print("Living Patients: ",living_patients)


Deceased Patients:  96
Living Patients:  203


### (m)

In [699]:
male_patients = heart_df[heart_df['sex'] == 1].count().values[0]
female_patients = heart_df[heart_df['sex'] == 0].count().values[0]
print("Male Patients: ",male_patients)
print("Female Patients: ",female_patients)


Male Patients:  194
Female Patients:  105


### (n)

In [700]:
smoking_patients = heart_df[heart_df['smoking'] == 1].count().values[0]
nonsmoking_patients = heart_df[heart_df['smoking'] == 0].count().values[0]
print("Smoking Patients: ",smoking_patients)
print("NonSmoking Patients: ",nonsmoking_patients)


Smoking Patients:  96
NonSmoking Patients:  203


## Problem 2.2

In [701]:

heart_df.loc[heart_df['age'] < 0, 'age'] = np.nan
heart_df.loc[heart_df['age'] > 125, 'age'] = np.nan

heart_df.loc[(heart_df['anaemia'] != 1) & (heart_df['anaemia'] != 0), 'anaemia'] = np.nan

heart_df.loc[heart_df['creatinine_phosphokinase'] < 0, 'creatinine_phosphokinase'] = np.nan
heart_df.loc[heart_df['creatinine_phosphokinase'] > 10000, 'creatinine_phosphokinase'] = np.nan

heart_df.loc[(heart_df['diabetes'] != 1) & (heart_df['diabetes'] != 0), 'diabetes'] = np.nan

heart_df.loc[heart_df['ejection_fraction'] < 0, 'ejection_fraction'] = np.nan
heart_df.loc[heart_df['ejection_fraction'] > 100, 'ejection_fraction'] = np.nan

heart_df.loc[(heart_df['high_blood_pressure'] != 1) & (heart_df['high_blood_pressure'] != 0), 'high_blood_pressure'] = np.nan

heart_df.loc[heart_df['platelets'] < 20000, 'platelets'] = np.nan
heart_df.loc[heart_df['platelets'] > 1000000, 'platelets'] = np.nan

heart_df.loc[heart_df['serum_creatinine'] < 0, 'serum_creatinine'] = np.nan
heart_df.loc[heart_df['serum_creatinine'] > 10, 'serum_creatinine'] = np.nan

heart_df.loc[heart_df['serum_sodium'] < 100, 'serum_sodium'] = np.nan
heart_df.loc[heart_df['serum_sodium'] > 175, 'serum_sodium'] = np.nan

heart_df.loc[(heart_df['sex'] != 1) & (heart_df['sex'] != 0), 'sex'] = np.nan

heart_df.loc[(heart_df['smoking'] != 1) & (heart_df['smoking'] != 0), 'smoking'] = np.nan

heart_df.loc[heart_df['time'] < 0, 'time'] = np.nan
heart_df.loc[heart_df['time'] > 700, 'time'] = np.nan

heart_df.loc[(heart_df['DEATH_EVENT'] != 1) & (heart_df['DEATH_EVENT'] != 0), 'DEATH_EVENT'] = np.nan


The oldest age ever recorded is 122 years old, so 125 is a good upper limit. Due to Rhabdomyolysis, it is possible for creatinine phosphokinase to reach up to 10000 mcg/L, so 10000 is a good upper bound. A percent for ejection fraction can never be above 100. Due to thrombocytosis, it is possible for platelet counts to reach up to 1000000 kiloplatelets/mL. With liver dysfunction it's possible to hit high single digits, which makes 10 a good upper limit. Hyponatremia and Hypernatremia suggest that it's possible to reach limits as low as 100 or as high as 175, which makes those good bounds to use for the data. A valid range that I have instated for time between checkups is up to 2 years without a checkup. All lower bounds not specifically stated are considered to be non-negative values.

## Problem 2.3

In [702]:
curr_columns = heart_df.columns
for col in curr_columns:
    if (col == 'anaemia' or col == 'diabetes' or col == 'high_blood_pressure' or col == 'sex' or col == 'smoking' or col == 'DEATH_EVENT'):
        continue
    curr_mean = heart_df[col].mean(skipna=True)
    curr_std = heart_df[col].std(skipna=True)
    heart_df[col] = (heart_df[col]-curr_mean)/curr_std

heart_df.head(20)

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0   1.190949      0.0                  0.000165       0.0          -1.527998   
1  -0.490457      0.0                  7.502063       0.0          -0.007065   
2   0.350246      0.0                 -0.449186       0.0          -1.527998   
3  -0.910808      1.0                 -0.485257       0.0          -1.527998   
4   0.350246      1.0                 -0.434757       1.0          -1.527998   
5   2.452003      1.0                 -0.551217       0.0           0.161928   
6   1.190949      1.0                 -0.346124       0.0          -1.950479   
7  -0.070106      1.0                 -0.275011       1.0           1.851853   
8   0.350246      0.0                 -0.437849       0.0           2.274335   
9   1.611300      1.0                 -0.472890       0.0          -0.260554   
10  1.190949      1.0                 -0.516176       0.0          -0.007065   
11  0.098035      0.0                 -0.361583       0.0          -1.105517   
12 -1.331160      1.0                  0.411384       0.0          -0.683035   
13 -0.910808      1.0                 -0.426512       0.0          -0.007065   
14 -0.994879      1.0                 -0.517207       0.0          -0.683035   
15  1.779441      1.0                 -0.209051       0.0           1.006890   
16  2.199792      1.0                 -0.446094       0.0          -0.007065   
17 -1.331160      0.0                  0.000165       0.0          -2.034976   
18  0.770597      1.0                 -0.470829       0.0          -1.105517   
19 -1.078949      1.0                  0.000165       1.0           1.429372   

    high_blood_pressure     platelets  serum_creatinine  serum_sodium  sex  \
0                   1.0  1.678834e-02          0.489237     -1.501519  1.0   
1                   0.0  7.523048e-09         -0.284076     -0.141739  1.0   
2                   0.0 -1.036336e+00         -0.090748     -1.728149  1.0   
3                   0.0 -5.455595e-01          0.489237      0.084892  1.0   
4                   0.0  6.507077e-01          1.262550     -4.674340  0.0   
5                   1.0 -6.069065e-01          0.682565     -1.048259  1.0   
6                   0.0 -1.394193e+00         -0.187412      0.084892  1.0   
7                   0.0  1.949220e+00         -0.284076     -1.274889  1.0   
8                   0.0  7.523048e-09          0.102580      0.311522  0.0   
9                   1.0  1.274403e+00          7.739045     -0.821629  1.0   
10                  1.0  1.069912e+00          2.519183     -1.274889  1.0   
11                  1.0 -1.059057e-01         -0.477404      0.764782  1.0   
12                  0.0 -1.302173e+00         -0.284076      0.084892  1.0   
13                  1.0  1.292579e-01         -0.284076      0.084892  1.0   
14                  1.0  1.673158e+00         -0.380740      0.311522  0.0   
15                  0.0 -2.212154e+00         -0.090748     -0.141739  1.0   
16                  0.0 -1.388518e-02         -0.477404      0.764782  1.0   
17                  0.0 -9.954377e-01         -0.574068     -2.181409  1.0   
18                  1.0 -2.694978e-01         -0.380740      0.764782  0.0   
19                  0.0 -1.803174e+00          0.489237     -3.541189  0.0   

    smoking      time  DEATH_EVENT  
0       0.0 -1.626775          1.0  
1       0.0 -1.601007          1.0  
2       1.0 -1.588122          1.0  
3       0.0 -1.588122          1.0  
4       0.0 -1.575238          1.0  
5       1.0 -1.575238          1.0  
6       0.0 -1.549470          1.0  
7       1.0 -1.549470          1.0  
8       0.0 -1.549470          1.0  
9       1.0 -1.549470          1.0  
10      1.0 -1.549470          1.0  
11      1.0 -1.549470          1.0  
12      0.0 -1.536586          1.0  
13      0.0 -1.536586          1.0  
14      0.0 -1.523701          0.0  
15      0.0 -1.510817          1.0  
16      0.0 -1.497933          1.0  
17      0.0 -1.497933  

## Problem 2.4

In [703]:
heart_df = pd.get_dummies(heart_df, columns=['DEATH_EVENT'], prefix='DEATH_EVENT')
display(heart_df)
heart_df.to_csv('heart_failure_dataset_processed.csv', index=False)

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    1.190949      0.0                  0.000165       0.0          -1.527998   
1   -0.490457      0.0                  7.502063       0.0          -0.007065   
2    0.350246      0.0                 -0.449186       0.0          -1.527998   
3   -0.910808      1.0                 -0.485257       0.0          -1.527998   
4    0.350246      1.0                 -0.434757       1.0          -1.527998   
..        ...      ...                       ...       ...                ...   
294  0.098035      0.0                 -0.536789       1.0          -0.007065   
295 -0.490457      0.0                  1.276075       0.0          -0.007065   
296 -1.331160      0.0                  1.523425       1.0           1.851853   
297 -1.331160      0.0                  1.887234       0.0          -0.007065   
298 -0.910808      0.0                 -0.397655       0.0           0.584409   

     high_blood_pressure     platelets  serum_creatinine  serum_sodium  sex  \
0                    1.0  1.678834e-02          0.489237     -1.501519  1.0   
1                    0.0  7.523048e-09         -0.284076     -0.141739  1.0   
2                    0.0 -1.036336e+00         -0.090748     -1.728149  1.0   
3                    0.0 -5.455595e-01          0.489237      0.084892  1.0   
4                    0.0  6.507077e-01          1.262550     -4.674340  0.0   
..                   ...           ...               ...           ...  ...   
294                  1.0 -1.107907e+00         -0.284076      1.444672  1.0   
295                  0.0  6.791087e-02         -0.187412      0.538152  0.0   
296                  0.0  4.893878e+00         -0.574068      0.311522  0.0   
297                  0.0 -1.261275e+00          0.005916      0.764782  1.0   
298                  0.0  1.345974e+00          0.199244     -0.141739  1.0   

     smoking      time  DEATH_EVENT_0.0  DEATH_EVENT_1.0  
0        0.0 -1.626775            False             True  
1        0.0 -1.601007            False             True  
2        1.0 -1.588122            False             True  
3        0.0 -1.588122            False             True  
4        0.0 -1.575238            False             True  
..       ...       ...              ...              ...  
294      1.0  1.800432             True            False  
295      0.0  1.813317             True            False  
296      0.0  1.903506             True            False  
297      1.0  1.929275             True            False  
298      1.0  1.993696             True            False  

[299 rows x 14 columns]

## Problem 2.5

In [704]:
heart_df.corr()

age   anaemia  creatinine_phosphokinase  \
age                       1.000000  0.088006                 -0.081584   
anaemia                   0.088006  1.000000                 -0.190741   
creatinine_phosphokinase -0.081584 -0.190741                  1.000000   
diabetes                 -0.101012 -0.012729                 -0.009639   
ejection_fraction         0.060098  0.031557                 -0.044080   
high_blood_pressure       0.093289  0.038182                 -0.070590   
platelets                -0.052354 -0.043786                  0.024463   
serum_creatinine          0.159187  0.052174                 -0.016408   
serum_sodium             -0.045966  0.041882                  0.059550   
sex                       0.065430 -0.094769                  0.079791   
smoking                   0.018668 -0.107290                  0.002421   
time                     -0.224068 -0.141414                 -0.009346   
DEATH_EVENT_0.0          -0.253729 -0.066270                 -0.062728   
DEATH_EVENT_1.0           0.253729  0.066270                  0.062728   

                          diabetes  ejection_fraction  high_blood_pressure  \
age                      -0.101012           0.060098             0.093289   
anaemia                  -0.012729           0.031557             0.038182   
creatinine_phosphokinase -0.009639          -0.044080            -0.070590   
diabetes                  1.000000          -0.004850            -0.012732   
ejection_fraction        -0.004850           1.000000             0.024445   
high_blood_pressure      -0.012732           0.024445             1.000000   
platelets                 0.092193           0.072177             0.049963   
serum_creatinine         -0.046975          -0.011302            -0.004935   
serum_sodium             -0.089551           0.175902             0.037109   
sex                      -0.157730          -0.148386            -0.104615   
smoking                  -0.147173          -0.067315            -0.055711   
time                      0.033726           0.041729            -0.196439   
DEATH_EVENT_0.0           0.001943           0.268603            -0.079351   
DEATH_EVENT_1.0          -0.001943          -0.268603             0.079351   

                          platelets  serum_creatinine  serum_sodium       sex  \
age                       -0.052354          0.159187     -0.045966  0.065430   
anaemia                   -0.043786          0.052174      0.041882 -0.094769   
creatinine_phosphokinase   0.024463         -0.016408      0.059550  0.079791   
diabetes                   0.092193         -0.046975     -0.089551 -0.157730   
ejection_fraction          0.072177         -0.011302      0.175902 -0.148386   
high_blood_pressure        0.049963         -0.004935      0.037109 -0.104615   
platelets                  1.000000         -0.041198      0.062125 -0.125120   
serum_creatinine          -0.041198          1.000000     -0.189095  0.006970   
serum_sodium               0.062125         -0.189095      1.000000 -0.027566   
sex                       -0.125120          0.006970     -0.027566  1.000000   
smoking                    0.028234         -0.027414      0.004813  0.445892   
time                       0.010514         -0.149315      0.087640 -0.015608   
DEATH_EVENT_0.0            0.049139         -0.294278      0.195204  0.004316   
DEATH_EVENT_1.0           -0.049139          0.294278     -0.195204 -0.004316   

                           smoking      time  DEATH_EVENT_0.0  DEATH_EVENT_1.0  
age                       0.018668 -0.224068        -0.253729         0.253729  
anaemia                  -0.107290 -0.141414        -0.066270         0.066270  
creatinine_phosphokinase  0.002421 -0.009346        -0.062728         0.062728  
diabetes                 -0.147173  0.033726         0.001943        -0.001943  
ejection_fraction        -0.067315  0.041729         0.268603        -0.268603  
high_blood_pressure      -0.055711 -0.196439   

### (a)

The two variables that are most positively correlated with DEATH_EVENT are age and serum_creatine.

### (b)

The two variables that are most negatively correlated with DEATH_EVENT are time and ejection_fraction.

### (c)

It makes sense that age should have a positive correlation on DEATH_EVENT because it's likely that people simply die of old age.

### (d)

It makes sense that ejection_fraction has a negative correlation with DEATH_EVENT because the more blood that's leaving the heart the stronger the heart is, so the less likely the individual is to die. It also makes sense that time has a negative correlation with DEATH_EVENT because if the individual is not visiting the doctor frequently then they may not have that many complications they need to see the doctor for.